pkg/rc/consul_backed.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/square/p2/pkg/pods"
+)
+
+type consulKV interface {
+	Put(*api.KVPair, *api.WriteOptions) (*api.WriteMeta, error)
+}
+
+type consulRc struct {
+	innerRc ReplicationController
+	kv      consulKV
+	kpStore kpStore
+}
+
+type kvPair struct {
+	key   string
+	value []byte
+}
+
+// consulRc implements the ReplicationController interface.
+var _ ReplicationController = &consulRc{}
+
+func NewConsulBacked(consulClient *api.Client, innerRc ReplicationController) (*consulRc, error) {
+	rc := &consulRc{
+		innerRc: innerRc,
+		kv:      consulClient.KV(),
+		kpStore: kp.NewFromClient(consulClient),
+	}
+
+	// Store everything related to the RC into consul, bail if errored.
+	_, err := rc.kpStore.SetPod(rc.consulPrefix()+"/pod_manifest", *innerRc.Manifest())
+	if err != nil {
+		return nil, err
+	}
+
+	err = rc.consulPut([]kvPair{
+		kvPair{key: "node_selector", value: []byte(innerRc.NodeSelector().String())},
+		kvPair{key: "disabled", value: []byte("false")},
+		kvPair{key: "replicas_desired", value: []byte("0")},
+		kvPair{key: "status", value: []byte("pending")},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return rc, nil
+}
+
+func (rc *consulRc) consulPutOne(key string, value []byte) error {
+	prefix := rc.consulPrefix()
+	p := &api.KVPair{Key: prefix + "/" + key, Value: value}
+	_, err := rc.kv.Put(p, nil)
+	return err
+}
+
+// consulPut attempts to put multiple KV Pairs into consul, returning an error if any fail.
+// For now, this performs no CAS or other locking.
+// It is assumed that only one process will be writing to an RC at a time.
+func (rc *consulRc) consulPut(pairs []kvPair) error {
+	for _, pair := range pairs {
+		err := rc.consulPutOne(pair.key, pair.value)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rc *consulRc) ID() string {
+	return rc.innerRc.ID()
+}
+
+func (rc *consulRc) Manifest() *pods.Manifest {
+	return rc.innerRc.Manifest()
+}
+
+func (rc *consulRc) NodeSelector() labels.Selector {
+	return rc.innerRc.NodeSelector()
+}
+
+func (rc *consulRc) IncreaseReplicas(n int) (int, error) {
+	err := rc.mutateDesired(rc.innerRc.DesiredReplicas() + n)
+	if err != nil {
+		return rc.innerRc.DesiredReplicas(), err
+	}
+	return rc.innerRc.IncreaseReplicas(n)
+}
+
+func (rc *consulRc) DecreaseReplicas(n int) (int, error) {
+	err := rc.mutateDesired(rc.innerRc.DesiredReplicas() - n)
+	if err != nil {
+		return rc.innerRc.DesiredReplicas(), err
+	}
+	return rc.innerRc.DecreaseReplicas(n)
+}
+
+func (rc *consulRc) mutateDesired(n int) error {
+	return rc.consulPutOne("replicas_desired", []byte(strconv.Itoa(n)))
+}
+
+func (rc *consulRc) DesiredReplicas() int {
+	return rc.innerRc.DesiredReplicas()
+}
+
+func (rc *consulRc) Disable() error {
+	err := rc.consulPutOne("disabled", []byte("true"))
+	if err != nil {
+		return err
+	}
+	return rc.innerRc.Disable()
+}
+
+func (rc *consulRc) Meet() error {
+	return rc.innerRc.Meet()
+}
+
+func (rc *consulRc) MeetEvery(every <-chan interface{}) chan<- error {
```

Comment:
I think I'm a bit confused about the purpose of the `consulRc`. It looks like you would use it to send instructions to the RC to update things, ie the replica count. But then it immediately goes on to do the update itself (via `innerRc`), implying that this is the server code. I would have imagined the `consulRc` basically doing what I see `memoryRc` doing but managed by some parent goroutine that is holding a watch on the rc's consul endpoint. How did you envision clients sending updates to Consul for RC changes?